packages/vincent-tools/vincent-tool-erc20-approval/src/lib/vincent-tool.ts

```diff
@@ -0,0 +1,180 @@
+
+export const Erc20ApprovalToolParamsSchema = z.object({
+  rpcUrl: z.string(),
+  chainId: z.number(),
+  pkpEthAddress: z.string(),
+  spenderAddress: z.string(),
+
+  tokenAddress: z.string(),
+  tokenDecimals: z.number(),
+  tokenAmount: z.number().refine((val) => val > 0, {
+    message: 'tokenAmount must be greater than 0',
+  }),
+});
+
+export const Erc20ApprovalToolPrecheckSuccessSchema = z.object({
+  allow: z.literal(true),
+  existingApprovalSufficient: z.boolean(),
+});
+
+export const Erc20ApprovalToolPrecheckFailSchema = z.object({
+  allow: z.literal(false),
+  error: z.string(),
+});
+
+export const Erc20ApprovalToolExecuteSuccessSchema = z.object({
+  // Whether the existing approval amount is sufficient for the requested amount
+  existingApprovalSufficient: z.boolean(),
+  // Transaction hash if a new approval was created, undefined if existing approval was used
+  approvalTxHash: z.string().optional(),
+  // The approved amount that is now active (either from existing or new approval)
+  approvedAmount: z.string(),
+  // The token address that was approved
+  tokenAddress: z.string(),
+  // The token decimals that was approved
+  tokenDecimals: z.number(),
+  // The spender address that was approved
+  spenderAddress: z.string(),
+});
+
+export const Erc20ApprovalToolExecuteFailSchema = z.object({
+  error: z.string(),
+});
+
+export const Erc20ApprovalToolDef = createVincentTool({
+  // packageName: '@lit-protocol/vincent-tool-erc20-approval' as const,
+
+  toolParamsSchema: Erc20ApprovalToolParamsSchema,
+  policyMap: createPolicyMapFromToolPolicies([]),
+
+  precheckSuccessSchema: Erc20ApprovalToolPrecheckSuccessSchema,
+  precheckFailSchema: Erc20ApprovalToolPrecheckFailSchema,
+
+  executeSuccessSchema: Erc20ApprovalToolExecuteSuccessSchema,
+  executeFailSchema: Erc20ApprovalToolExecuteFailSchema,
+
+  precheck: async ({ toolParams }, { policiesContext, fail, succeed }) => {
+    if (!policiesContext.allow) return fail({ allow: false, error: 'Policy check failed' });
+
+    const { rpcUrl, pkpEthAddress, spenderAddress, tokenAddress, tokenDecimals, tokenAmount } =
+      toolParams;
+
+    const client = createPublicClient({
+      transport: http(rpcUrl),
+    });
+
+    await checkNativeTokenBalance({
+      client,
+      pkpEthAddress: pkpEthAddress as `0x${string}`,
+    });
+
+    const currentAllowance = await getCurrentAllowance({
+      client,
+      tokenAddress: tokenAddress as `0x${string}`,
+      owner: pkpEthAddress as `0x${string}`,
+      spender: spenderAddress as `0x${string}`,
+    });
+
+    const requiredAmount = parseUnits(tokenAmount.toString(), tokenDecimals);
+
+    return succeed({
+      allow: true,
+      existingApprovalSufficient: currentAllowance >= requiredAmount,
+    });
+  },
+  execute: async ({ toolParams }, { succeed, fail, policiesContext }) => {
+    if (!policiesContext.allow) return fail({ error: 'Policy check failed' });
```

Comment:
This should never happen - `policiesContext` in this function should only ever be `allow: true` or we wouldn't call the function. Kind of surprised TS didn't flag this is being impossible 🤔